pkg/kubelet/kubelet.go

```diff
@@ -699,11 +699,19 @@ func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
 		}
 		// TODO(roberthbailey): Can we do this without having credentials to talk
 		// to the cloud provider?
-		instanceID, err := instances.ExternalID(kl.hostname)
+		externalID, err := instances.ExternalID(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get external ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = externalID
+		// TODO: We can't assume that the node has credentials to talk to the
+		// cloudprovider from arbitrary nodes. At most, we should talk to a
+		// local metadata server here.
+		instanceID, err := instances.InstanceID(node.Name)
 		if err != nil {
 			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
 		}
-		node.Spec.ExternalID = instanceID
+		node.Spec.ProviderID = kl.cloud.ProviderName() + "://" + instanceID
```

Comment:
Where exactly? I wanted to avoid to give the control to the cloud providers, the enforced format is `<ProviderName>://<ProviderSpecificInstanceID>` so I want to gather the two values separately and join them.

I'll try to encapsulate this in `cloudprovider`.